File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        boolean[] visited = new boolean[n];
        Queue<Integer> q = new ArrayDeque<>();
        for(int i = 0; i < n; i ++) {
            if(visited[i]) continue;
            q.offer(i);
            visited[i] = true;
            while(!q.isEmpty()) {
                int time = q.size();
                for(int t = 0; t < time; t ++) {
                    int temp = q.poll();
                    for(int j = 0; j < n; j ++) {
                        if(j == temp || visited[j] || computers[temp][j] != 1) continue;
                        visited[j] = true;
                        q.offer(j);
                    }
                }
            }
            answer++;
        }
        return answer;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution(n, computers) {
    var answer = 0;
    // 2차원 배열 생성
    let node = Array.from(Array(n + 1), () => Array(n + 1).fill(false));
    // 인접행렬 생성
    for(let i = 0; i < n; i ++) {
        for(let j = 0; j < n; j ++) {
            if(!!computers[i][j]) {
                node[i + 1][j + 1] = true;
                node[j + 1][i + 1] = true;
            }
        }
    }
    // bfs
    let q = []; // 큐 생성
    let visited = Array(n + 1).fill(false); // 방문체크 배열 초기화
    for(let i = 1; i <= n; i ++) {
        if(visited[i]) continue;
        q.push(i);  // 처음 노드 추가
        visited[i] = true;  // 방문 체크
        while(q.length > 0) {
            let time = q.length;
            for(let t = 0; t < time; t ++) {
                let temp = q.shift();
                for(let j = 1; j <= n; j ++) {
                    if(j == temp || !node[temp][j] || visited[j]) continue;
                    q.push(j);
                    visited[j] = true;
                }
            }
        }
        answer ++;
    }
    return answer;
}